lib/channel.js

```diff
@@ -103,21 +133,71 @@ class Channel {
     const packet = metautil.jsonParse(data);
     if (!packet) {
       const error = new Error('JSON parsing error');
-      this.error(400, { error, pass: true });
+      this.error(500, { error, pass: true });
       return;
     }
+    const [callType] = Object.keys(packet);
+    if (callType === 'call') return this.handleRpcPacket(packet);
+    else if (callType === 'stream') return this.handleStreamPacket(packet);
+    const error = new Error('Packet structure error');
+    this.error(500, { error, pass: true });
+  }
+
+  binary(data) {
+    try {
+      const { streamId, payload } = MetacomChunk.decode(data);
+      const upstream = this.client.streams.get(streamId);
+      if (upstream) {
+        upstream.push(payload);
+      } else {
+        const error = new Error(`Stream ${streamId} is not initialized`);
+        this.error(400, { callId: streamId, error, pass: true });
+      }
+    } catch (error) {
+      this.error(400, { callId: 0, error });
+    }
+  }
+
+  handleRpcPacket(packet) {
     const [callType, target] = Object.keys(packet);
     const callId = parseInt(packet[callType], 10);
     const args = packet[target];
     if (callId && args) {
       const [interfaceName, methodName] = target.split('/');
-      this.rpc(callId, interfaceName, methodName, args);
+      void this.rpc(callId, interfaceName, methodName, args);
       return;
     }
     const error = new Error('Packet structure error');
     this.error(400, { callId, error, pass: true });
   }
 
+  async handleStreamPacket(packet) {
+    const { stream: streamId, name, size, status } = packet;
+    const stream = this.client.streams.get(streamId);
+    if (name && typeof name === 'string' && Number.isSafeInteger(size)) {
+      if (stream) {
+        const error = new Error(`Stream ${name} is already initialized`);
+        this.error(400, { callId: streamId, error, pass: true });
+      } else {
+        const streamData = { streamId, name, size };
+        const stream = new MetacomReadable(streamData);
+        this.client.streams.set(streamId, stream);
+      }
+    } else if (!stream) {
+      const error = new Error(`Stream ${streamId} is not initialized`);
+      this.error(400, { callId: streamId, error, pass: true });
+    } else if (status === 'end') {
+      await stream.close();
+      this.client.streams.delete(streamId);
+    } else if (status === 'terminate') {
+      await stream.terminate();
+      this.client.streams.delete(streamId);
+    } else {
+      const error = new Error('Stream packet structure error');
+      this.error(400, { callId: streamId, error, pass: true });
```

Comment:
In channel.error function we don't have the 'pass' argument in destructing and We create a new variable with same name.
I checked error function in 'ws.js' and 'http.js'.
Error function doesn't overrided there.
Why is the "pass" argument being passed here?

Reply:
It seems it was forgotten to be removed in one of previous commits. See https://github.com/metarhia/metacom/commit/53056edaf3ac91879b535f5d4a04089b5951e864

There are two places in channel.js to be fixed.
You can create a PR for this.